### PR TITLE
(WIP) Aws linker

### PR DESCRIPTION
**Pull request by RobinL**

### CODE REVIEW

File: splink/aws/aws_linker.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+    def random_sample_sql(self, proportion, sample_size):
+        if proportion == 1.0:
+            return ""
+
+        sample_size = int(sample_size)
+
+        return (
+            "where unique_id IN (SELECT unique_id FROM __splink__df_concat_with_tf"
+            f" ORDER BY RANDOM() LIMIT {sample_size})"
```
Comment (RobinL) [inline splink/aws/aws_linker.py:87]: if you use `tablesample` this will run a bit faster

File: splink/aws/aws_linker.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+import time
+
+class AWSDataFrame(SplinkDataFrame):
+    def __init__(self, templated_name, physical_name, aws_linker):
+        super().__init__(templated_name, physical_name)
+        self.aws_linker = aws_linker
+
+    @property
+    def columns(self):
+        d = self.as_record_dict(1)[0]
```
[the remark this answers is not in the document]
Comment (Th368MoJ) [reply splink/aws/aws_linker.py]: Resolved [here](https://github.com/moj-analytical-services/splink/pull/363/commits/b4f9240e1519004c3aee46d109e8525e3e4d0926)

File: splink/aws/aws_linker.py (hunk 1 above)
Comment (Th368MoJ) [reply splink/aws/aws_linker.py]: Resolved [here](https://github.com/moj-analytical-services/splink/pull/363/commits/49b7ffc1dffe1bfa09edd8090d35fb039d725f87). Though, it does seem to be slower for the smaller df example I ran it on.

File: splink/aws/aws_linker.py [hunk 3]
```diff
@@ -0,0 +1,186 @@
+import sqlglot
```
Comment (RobinL) [inline splink/aws/aws_linker.py:1]: I think it should be called `athena_linker.py` - i.e. let's be as specific as possible with naming

This implies various other renamings e.g. `AWSDataFrame` becomes `AthenaDataFrame` and so on - prob need to find all instances of `aws` to find everything that needs to be renamed

File: splink/aws/aws_linker.py [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+        if self.aws_linker.ctas_query_info[self.physical_name]["ctas_table"] != self.physical_name:
+            raise ValueError(
+                    f"The recorded metadata for {self.physical_name} that you're "
+                    "attempting to delete does not match the recorded metadata on s3. "
+                    "To prevent any tables becoming corrupted on s3, this run will be "
+                    "terminated. Please retry the link/dedupe job and report the issue "
+                    "if this error persists."
+                )
+
+    def as_record_dict(self, limit=None):
```
Comment (RobinL) [inline splink/aws/aws_linker.py:68]: In addition to the `as_record_dict()` method, we also need an `as_pandas_dataframe` method (which, in this case, is the same, except returns `out_df`, without needing the `to_dict(orient='records)`) part

File: splink/aws/aws_linker.py [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+
+    def get_schema_info(self, input_table):
+        t = input_table.split(".")
+        return t if len(t) > 1 else [self.aws_linker.output_schema, self.physical_name]
+    
+
+class AWSLinker(Linker):
+    def __init__(self, settings_dict: dict,
+                 boto3_session: boto3.session.Session,
+                 output_schema: str,
```
Comment (RobinL) [inline splink/aws/aws_linker.py:92]: Is the correct terminology here (in the world of Athena) a `database`?  

If so I think for clarity for the end user the argument here should be called `output_database`.  `output_schema` can then be used internally e.g. `self.output_schema = output_database`.

File: run_aws_linker.ipynb [hunk 6]
```diff
@@ -0,0 +1,278 @@
+{
```
Comment (RobinL) [inline run_aws_linker.ipynb:1]: Can we move this file into the `splink3_demos` branch of the `splink_demos` repo before we merge

**Comment by RobinL** [comment]
Note I have tested this against the 50k synthetic dataset in `splink_demos` and everything's working nicely!

**Comment by RobinL** [comment]
This is fantastic!

File: run_aws_linker.ipynb (hunk 6 above)
Comment (Th368MoJ) [reply run_aws_linker.ipynb:1]: Yeah, I think that makes sense. This `.ipynb` script was never meant to stick around once we merged. It was simply to allow for wider testing.

**Comment by Th368MoJ** [comment]
I'll make your suggested changes and then lint it

File: splink/aws/aws_linker.py (hunk 5 above)
Comment (Th368MoJ) [reply splink/aws/aws_linker.py:92]: Yeah, that's fine. I changed it to `schema_output` to better align with the `duckdb` side of things.

Athena is a bit of a mess when it comes to what they actually mean by the data prefix (i.e. `splink.`) - the data engineers all refer to it as a schema, while amazon themselves seem to call it a database... in reality, it's not really either 🤷

**Comment by Th368MoJ** [comment]
## Some initial benchmarks from testing this linker:

No screenshot, but the smallest linkage job (which all of the benchmark results we currently have are for), takes about 3 minutes (which is incredibly slow in the grander scheme of things).

However, the larger dataset benchmarks are very promising:
<hr>
A 9 million row dedupe job took:
Screenshot 2022-03-30 at 23 23 21

The 17.6 million row job 
Screenshot 2022-03-30 at 23 42 44
took:


And the largest deduping job we tested -**35 million initial rows**:
* **Total time taken:** 52 minutes
* **Output rows:** 3.9 billion
* Prediction time (performing our actual links) can also be seen below.

Screenshot 2022-04-01 at 13 27 29

<hr>
In summary:
The linker performs very well on larger jobs (52 minutes to run a 35 million row dedupe), while also reducing the need to read/write from aws.

On smaller linkage jobs (<1 million up to <5 million), `duckdb` offers the fastest linking, while also removing any need to interact with aws and is the suggested route forward for users.
